File: app/spring/src/main/java/az/ecommerce/payment/domain/Payment.java
package az.ecommerce.payment.domain;

import az.ecommerce.payment.domain.enums.PaymentMethodEnum;
import az.ecommerce.payment.domain.enums.PaymentStatusEnum;
import az.ecommerce.payment.domain.event.PaymentCompletedEvent;
import az.ecommerce.payment.domain.event.PaymentCompletedIntegrationEvent;
import az.ecommerce.payment.domain.event.PaymentCreatedEvent;
import az.ecommerce.payment.domain.event.PaymentFailedEvent;
import az.ecommerce.payment.domain.event.PaymentFailedIntegrationEvent;
import az.ecommerce.payment.domain.valueobject.PaymentId;
import az.ecommerce.product.domain.valueobject.Money;
import az.ecommerce.shared.domain.AggregateRoot;

import java.time.Instant;
import java.util.UUID;

public class Payment extends AggregateRoot {

    private final PaymentId id;
    private final UUID orderId;
    private final UUID userId;
    private final Money amount;
    private final PaymentMethodEnum method;
    private PaymentStatusEnum status;
    private String transactionId;
    private String failureReason;
    private Instant completedAt;

    private Payment(PaymentId id, UUID orderId, UUID userId, Money amount, PaymentMethodEnum method) {
        this.id = id;
        this.orderId = orderId;
        this.userId = userId;
        this.amount = amount;
        this.method = method;
        this.status = PaymentStatusEnum.PENDING;
    }

    public static Payment initiate(UUID orderId, UUID userId, Money amount, PaymentMethodEnum method) {
        PaymentId id = PaymentId.generate();
        Payment p = new Payment(id, orderId, userId, amount, method);
        p.recordEvent(PaymentCreatedEvent.of(id, orderId));
        return p;
    }

    public static Payment reconstitute(PaymentId id, UUID orderId, UUID userId, Money amount,
                                       PaymentMethodEnum method, PaymentStatusEnum status,
                                       String transactionId, String failureReason) {
        Payment p = new Payment(id, orderId, userId, amount, method);
        p.status = status;
        p.transactionId = transactionId;
        p.failureReason = failureReason;
        return p;
    }

    public void startProcessing() {
        status.requireTransitionTo(PaymentStatusEnum.PROCESSING);
        this.status = PaymentStatusEnum.PROCESSING;
    }

    public void complete(String transactionId) {
        status.requireTransitionTo(PaymentStatusEnum.COMPLETED);
        this.status = PaymentStatusEnum.COMPLETED;
        this.transactionId = transactionId;
        this.completedAt = Instant.now();
        recordEvent(PaymentCompletedEvent.of(id, orderId));
        recordEvent(PaymentCompletedIntegrationEvent.of(id.value(), orderId, amount.amount()));
    }

    public void fail(String reason) {
        status.requireTransitionTo(PaymentStatusEnum.FAILED);
        this.status = PaymentStatusEnum.FAILED;
        this.failureReason = reason;
        recordEvent(PaymentFailedEvent.of(id, orderId, reason));
        recordEvent(PaymentFailedIntegrationEvent.of(id.value(), orderId, reason));
    }

    public PaymentId id() { return id; }
    public UUID orderId() { return orderId; }
    public UUID userId() { return userId; }
    public Money amount() { return amount; }
    public PaymentMethodEnum method() { return method; }
    public PaymentStatusEnum status() { return status; }
    public String transactionId() { return transactionId; }
    public String failureReason() { return failureReason; }
}

File: app/golang/internal/payment/domain/payment.go
// Package domain — Payment context
package domain

import (
	"time"

	"github.com/google/uuid"
	productDomain "github.com/orkhan/ecommerce/internal/product/domain"
	sharedDomain "github.com/orkhan/ecommerce/internal/shared/domain"
)

type PaymentID uuid.UUID

func GeneratePaymentID() PaymentID  { return PaymentID(uuid.New()) }
func (p PaymentID) UUID() uuid.UUID { return uuid.UUID(p) }
func (p PaymentID) String() string  { return uuid.UUID(p).String() }

// === ENUMS ===

type PaymentMethod string

const (
	PaymentMethodCreditCard   PaymentMethod = "CREDIT_CARD"
	PaymentMethodPayPal       PaymentMethod = "PAYPAL"
	PaymentMethodBankTransfer PaymentMethod = "BANK_TRANSFER"
	PaymentMethodStripe       PaymentMethod = "STRIPE"
)

type PaymentStatus string

const (
	PaymentStatusPending    PaymentStatus = "PENDING"
	PaymentStatusProcessing PaymentStatus = "PROCESSING"
	PaymentStatusCompleted  PaymentStatus = "COMPLETED"
	PaymentStatusFailed     PaymentStatus = "FAILED"
	PaymentStatusRefunded   PaymentStatus = "REFUNDED"
)

func (s PaymentStatus) RequireTransitionTo(target PaymentStatus) error {
	allowed := map[PaymentStatus]map[PaymentStatus]bool{
		PaymentStatusPending:    {PaymentStatusProcessing: true},
		PaymentStatusProcessing: {PaymentStatusCompleted: true, PaymentStatusFailed: true},
		PaymentStatusCompleted:  {PaymentStatusRefunded: true},
	}
	if !allowed[s][target] {
		return sharedDomain.NewDomainError("Yanlış payment status keçidi: " + string(s) + " → " + string(target))
	}
	return nil
}

// === EVENTS ===

type PaymentCreatedEvent struct {
	sharedDomain.BaseEvent
	PaymentID PaymentID
	OrderID   uuid.UUID
}

func (PaymentCreatedEvent) EventName() string { return "PaymentCreated" }

type PaymentCompletedEvent struct {
	sharedDomain.BaseEvent
	PaymentID PaymentID
	OrderID   uuid.UUID
}

func (PaymentCompletedEvent) EventName() string { return "PaymentCompleted" }

type PaymentCompletedIntegrationEvent struct {
	sharedDomain.BaseEvent
	PaymentID uuid.UUID `json:"payment_id"`
	OrderID   uuid.UUID `json:"order_id"`
	Amount    int64     `json:"amount"`
}

func (PaymentCompletedIntegrationEvent) EventName() string  { return "PaymentCompletedIntegration" }
func (PaymentCompletedIntegrationEvent) RoutingKey() string { return "payment.completed" }

type PaymentFailedEvent struct {
	sharedDomain.BaseEvent
	PaymentID PaymentID
	OrderID   uuid.UUID
	Reason    string
}

func (PaymentFailedEvent) EventName() string { return "PaymentFailed" }

type PaymentFailedIntegrationEvent struct {
	sharedDomain.BaseEvent
	PaymentID uuid.UUID `json:"payment_id"`
	OrderID   uuid.UUID `json:"order_id"`
	Reason    string    `json:"reason"`
}

func (PaymentFailedIntegrationEvent) EventName() string  { return "PaymentFailedIntegration" }
func (PaymentFailedIntegrationEvent) RoutingKey() string { return "payment.failed" }

// === AGGREGATE ===

type Payment struct {
	sharedDomain.AggregateRoot

	id            PaymentID
	orderID       uuid.UUID
	userID        uuid.UUID
	amount        productDomain.Money
	method        PaymentMethod
	status        PaymentStatus
	transactionID string
	failureReason string
	completedAt   *time.Time
}

func Initiate(orderID, userID uuid.UUID, amount productDomain.Money, method PaymentMethod) *Payment {
	id := GeneratePaymentID()
	p := &Payment{
		id: id, orderID: orderID, userID: userID,
		amount: amount, method: method, status: PaymentStatusPending,
	}
	p.RecordEvent(PaymentCreatedEvent{
		BaseEvent: sharedDomain.NewBaseEvent(), PaymentID: id, OrderID: orderID,
	})
	return p
}

func Reconstitute(id PaymentID, orderID, userID uuid.UUID, amount productDomain.Money,
	method PaymentMethod, status PaymentStatus, transactionID, failureReason string) *Payment {
	return &Payment{
		id: id, orderID: orderID, userID: userID,
		amount: amount, method: method, status: status,
		transactionID: transactionID, failureReason: failureReason,
	}
}

func (p *Payment) StartProcessing() error {
	if err := p.status.RequireTransitionTo(PaymentStatusProcessing); err != nil {
		return err
	}
	p.status = PaymentStatusProcessing
	return nil
}

func (p *Payment) Complete(transactionID string) error {
	if err := p.status.RequireTransitionTo(PaymentStatusCompleted); err != nil {
		return err
	}
	now := time.Now()
	p.status = PaymentStatusCompleted
	p.transactionID = transactionID
	p.completedAt = &now
	p.RecordEvent(PaymentCompletedEvent{
		BaseEvent: sharedDomain.NewBaseEvent(), PaymentID: p.id, OrderID: p.orderID,
	})
	p.RecordEvent(PaymentCompletedIntegrationEvent{
		BaseEvent: sharedDomain.NewBaseEvent(),
		PaymentID: p.id.UUID(), OrderID: p.orderID, Amount: p.amount.Amount(),
	})
	return nil
}

func (p *Payment) Fail(reason string) error {
	if err := p.status.RequireTransitionTo(PaymentStatusFailed); err != nil {
		return err
	}
	p.status = PaymentStatusFailed
	p.failureReason = reason
	p.RecordEvent(PaymentFailedEvent{
		BaseEvent: sharedDomain.NewBaseEvent(),
		PaymentID: p.id, OrderID: p.orderID, Reason: reason,
	})
	p.RecordEvent(PaymentFailedIntegrationEvent{
		BaseEvent: sharedDomain.NewBaseEvent(),
		PaymentID: p.id.UUID(), OrderID: p.orderID, Reason: reason,
	})
	return nil
}

// Getters
func (p *Payment) ID() PaymentID                     { return p.id }
func (p *Payment) OrderID() uuid.UUID                { return p.orderID }
func (p *Payment) UserID() uuid.UUID                 { return p.userID }
func (p *Payment) Amount() productDomain.Money       { return p.amount }
func (p *Payment) Method() PaymentMethod             { return p.method }
func (p *Payment) Status() PaymentStatus             { return p.status }
func (p *Payment) TransactionID() string             { return p.transactionID }
func (p *Payment) FailureReason() string             { return p.failureReason }

// Repository
type Repository interface {
	Save(payment *Payment) error
	FindByID(id PaymentID) (*Payment, error)
}

// === STRATEGY PATTERN ===

// Gateway — Strategy interface
//
// Laravel: PaymentGatewayInterface  ·  Spring: PaymentGateway
// Go: interface (implicit implementation)
type Gateway interface {
	SupportedMethod() PaymentMethod
	Charge(amount productDomain.Money, reference string) GatewayResult
}

type GatewayResult struct {
	Success       bool
	TransactionID string
	ErrorMessage  string
}

func GatewayOK(txID string) GatewayResult {
	return GatewayResult{Success: true, TransactionID: txID}
}

func GatewayFail(msg string) GatewayResult {
	return GatewayResult{Success: false, ErrorMessage: msg}
}

// StrategyResolver — Strategy pattern resolver (DI map)
type StrategyResolver struct {
	gateways map[PaymentMethod]Gateway
}

func NewStrategyResolver(gws ...Gateway) *StrategyResolver {
	m := make(map[PaymentMethod]Gateway)
	for _, g := range gws {
		m[g.SupportedMethod()] = g
	}
	return &StrategyResolver{gateways: m}
}

func (r *StrategyResolver) Resolve(method PaymentMethod) (Gateway, error) {
	g, ok := r.gateways[method]
	if !ok {
		return nil, sharedDomain.NewDomainError("Bu ödəmə üsulu dəstəklənmir: " + string(method))
	}
	return g, nil
}
